### PR TITLE
Allow a non metrics-provider charm to provide rules

### PR DESCRIPTION
**Pull request by sed-i**
An existing use case is to have rules coming in dynamically from a constantly-updating git repo.
For this purpose, a new mechanism is needed for continually forwarding rules information as they are updated.

Prometheus config file is comprised of the following sections (per @mmanciop):

```
global:
  ... # set by juju config of this charm (prometheus)

rule_files:
  ... # THIS PR

scrape_configs:
... # Covered by the prometheus-scrape relation (prometheus_scrape.MetricsEndpointProvider and Consumer)

alerting:
... # alertmanager interface. Partially covered by the prometheus charm code.

remote_write:
... # prometheus_remote_write interface (#147)

remote_read:
... # We may never implement this and if we do, it is a prometheus_remote_read interface
```

This PR is for allowing a non metrics-provider charm to provide rules, for example:

```python
self.prom_config_subset = RuleFilesProvider(
    self,
    "prometheus-config",
    dir_path="/git/repo/prom_rules",
    recursive=True,
)
```

- [x] add JujuTopology class
- [x] convert rule-parsing methods into free functions
- [x] Create a new RuleFilesProvider

### CODE REVIEW

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 1]
```diff
@@ -1182,15 +1220,54 @@ def _scrape_metadata(self) -> dict:
...
+        self._relation_name = relation_name
+        self.topology = JujuTopology.from_charm(charm)
+        self.dir_path = dir_path
+        self.recursive = recursive
+
+        events = self._charm.on[self._relation_name]
+        self.framework.observe(events.relation_joined, self._update_relation_data)
+        self.framework.observe(events.relation_changed, self._update_relation_data)
+        self.framework.observe(self._charm.on.leader_elected, self._update_relation_data)
+        self.framework.observe(self._charm.on.config_changed, self._update_relation_data)
```
Comment (mmanciop) [inline lib/charms/prometheus_k8s/v0/prometheus_scrape.py:1254]: Todo here still relevant?

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 2]
```diff
@@ -474,6 +476,106 @@ def _sanitize_scrape_configuration(job) -> dict:
...
+    for path in filter(Path.is_file, Path(dir_path).glob("**/*.rule" if recursive else "*.rule")):
+        relpath = os.path.relpath(os.path.dirname(path), dir_path)
+
+        # Generate group name:
+        #  - prefix, from the relative path of the rule file;
+        #  - name, from juju topology
+        group_name = (
+            f"{'' if relpath == '.' else relpath.replace(os.path.sep, '_') + '_'}"
+            f"{topology.as_str_short_form()}_alerts"
+        )
```
Comment (Abuelodelanada) [inline lib/charms/prometheus_k8s/v0/prometheus_scrape.py:552]: This kind of comments sometimes indicates that is worth extracting this lines into a small function, for instance: `_generate_group_name(path, topology)`

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 3]
```diff
@@ -474,6 +476,106 @@ def _sanitize_scrape_configuration(job) -> dict:
...
+                try:
+                    # add "juju_" topology labels
+                    rule.get("labels", {}).update(topology.as_dict_long_form())
+
+                    # insert juju topology filters into a prometheus alert rule
+                    rule["expr"] = topology.render(rule["expr"])
+                except KeyError:
+                    logger.error("Invalid alert rule %s: missing an 'expr' property.", path.name)
+                else:
+                    alerts[group_name].append(rule)
```
Comment (Abuelodelanada) [inline lib/charms/prometheus_k8s/v0/prometheus_scrape.py:571]: The same here, maybe it's worth extracting this code into a custom function.

Besides, shouldn't we raise custom exceptions if `Exception` and `KeyError` is catched?

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 4]
```diff
@@ -1182,15 +1220,54 @@ def _scrape_metadata(self) -> dict:
...
-            "charm_name": f"{self._charm.meta.name}",
-        }
-        return metadata
+        return self.topology.as_dict()
+
+
+class RuleFilesProvider(Object):
+    """A 'prometheus_scrape' provider class for rule files only.
+
+    The RuleFilesProvider class if for sending rules to prometheus, unlike
```
Comment (Abuelodelanada) [inline lib/charms/prometheus_k8s/v0/prometheus_scrape.py:1231]: The RuleFilesProvider class **is** for sending...

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 5]
```diff
@@ -1182,15 +1220,54 @@ def _scrape_metadata(self) -> dict:
...
+     (including rules). This is useful for providing "free standing" rules.
+
+    Args:
+        charm: a charm instance that has the `prometheus_scrape` interface on the "provides" side.
+        relation_name: must match metadata.yaml and have `prometheus_scrape` as interface.
+        dir_path: root directory for the collection of rule files.
+        recursive: whether or not to scan for rule files recursively.
+    """
+
+    def __init__(self, charm: CharmBase, relation_name: str, dir_path: str, recursive=True):
```
Comment (Abuelodelanada) [inline lib/charms/prometheus_k8s/v0/prometheus_scrape.py]: What about having a default value for `relation_name`?

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 6]
```diff
@@ -1182,15 +1220,54 @@ def _scrape_metadata(self) -> dict:
...
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+        self.topology = JujuTopology.from_charm(charm)
+        self.dir_path = dir_path
+        self.recursive = recursive
```
Comment (Abuelodelanada) [inline lib/charms/prometheus_k8s/v0/prometheus_scrape.py:1256]: If these properties are not going to be public, I would prefix them with  `_`

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 7]
```diff
@@ -474,6 +476,106 @@ def _sanitize_scrape_configuration(job) -> dict:
...
+    """Dataclass for storing and formatting juju topology information."""
+
+    model: str
+    model_uuid: str
+    application: str
+    charm_name: str
+
+    @staticmethod
+    def from_charm(charm):
+        """Factory method for creating the topology dataclass from a given charm."""
```
[the remark this answers is not in the document]
Comment (sed-i) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py:490]: I only refactored the existing.
Not sure what you mean: replace charm_name with (sanitized) unit name?

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 8]
```diff
@@ -474,6 +476,106 @@ def _sanitize_scrape_configuration(job) -> dict:
...
+        return JujuTopology(
+            model=charm.model.name,
+            model_uuid=charm.model.uuid,
+            application=charm.model.app.name,
+            charm_name=charm.meta.name,
+        )
+
+    def as_str_short_form(self) -> str:
```
[the remark this answers is not in the document]
Comment (sed-i) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py:498]: Should probably be changed to
```python
@property
def identifier(self) -> str:
```
?

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 9]
```diff
@@ -474,6 +476,106 @@ def _sanitize_scrape_configuration(job) -> dict:
...
+        """Format the topology information into a terse string."""
+        return f"{self.model}_{self.model_uuid}_{self.application}"
+
+    def as_str_long_form(self) -> str:
```
[the remark this answers is not in the document]
Comment (sed-i) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py:502]: Should probably be changed to
```python
@property
def promql_labels(self) -> str:
```
?

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py [hunk 10]
```diff
@@ -474,6 +476,106 @@ def _sanitize_scrape_configuration(job) -> dict:
...
+            "juju_model_uuid": self.model_uuid,
+            "juju_application": self.application,
+        }
+
+    def render(self, template: str):
+        """Render a juju-topology template string with topology info."""
+        return template.replace("%%juju_topology%%", self.as_str_long_form())
+
+
+def load_alert_rules_from_dir(
```
[the remark this answers is not in the document]
Comment (sed-i) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py:525]: Good question. This function seems like it qualifies as a public interface. Not sure.

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py (hunk 2 above)
Comment (sed-i) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py:552]: Not sure about this: relpath makes little sense outside of this function so I am more inclined towards an inner function, wdyt?

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py (hunk 3 above)
Comment (sed-i) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py:571]: > it's worth extracting this code into a custom function

one option is to pass the `alerts` "by-ref":
`read_rule_file(alerts, group_name, path, rule_file, topology)`

another option is an inner function:
`read_rule_file(group_name, path, rule_file)`

wdyt?

> raise custom exceptions

Per discussion with @balbirthomas, and in line with the original code, file reading should proceed even if one of them is buggy.

File: lib/charms/prometheus_k8s/v0/prometheus_scrape.py (hunk 6 above)
Comment (sed-i) [reply lib/charms/prometheus_k8s/v0/prometheus_scrape.py:1256]: added `_` to `recursive`. I wonder if `dir_path` and `topology` make sense as public.

**Pull request merged**